hail/src/main/scala/is/hail/expr/ir/Infer.scala

```diff
@@ -1,89 +1,101 @@
-        UnaryOp.getReturnType(op, v.typ)
+        PType.canonical(UnaryOp.getReturnType(op, v.typ)).setRequired(v.pType.required)
       case ApplyComparisonOp(op, l, r) =>
         assert(l.typ isOfType r.typ, s"${l.typ.parsableString()} vs ${r.typ.parsableString()}")
-        TBoolean()
+        PBoolean(l.pType.required && r.pType.required)
       case ArrayRef(a, i) =>
         assert(i.typ.isOfType(TInt32()))
-        -coerce[TArray](a.typ).elementType
+        coerce[PArray](a.pType).elementType.setRequired(a.pType.required && i.pType.required)
       case ArraySort(a, ascending, _) =>
         assert(ascending.typ.isOfType(TBoolean()))
-        a.typ
+        val et = coerce[PArray](a.pType).elementType
+        PArray(et, a.pType.required)
       case ToSet(a) =>
-        TSet(coerce[TArray](a.typ).elementType)
+        val et = coerce[PArray](a.pType).elementType
+        PSet(et, a.pType.required)
       case ToDict(a) =>
-        val elt = coerce[TBaseStruct](coerce[TArray](a.typ).elementType)
-        TDict(elt.types(0), elt.types(1))
+        val elt = coerce[PBaseStruct](coerce[PArray](a.pType).elementType)
+        PDict(elt.types(0), elt.types(1), a.pType.required)
       case ToArray(a) =>
-        TArray(coerce[TContainer](a.typ).elementType)
+        val et = coerce[PContainer](a.pType).elementType
+        PArray(et, a.pType.required)
       case GroupByKey(collection) =>
-        val elt = coerce[TBaseStruct](coerce[TArray](collection.typ).elementType)
-        TDict(elt.types(0), TArray(elt.types(1)))
+        val elt = coerce[PBaseStruct](coerce[PArray](collection.pType).elementType)
+        // FIXME requiredness
+        PDict(elt.types(0), PArray(elt.types(1), required = false), collection.pType.required)
       case ArrayMap(a, name, body) =>
-        TArray(-body.typ, a.typ.required)
+        // FIXME: requiredness artifact of former IR bug, remove when possible
+        PArray(body.pType.setRequired(false), a.typ.required)
       case ArrayFilter(a, name, cond) =>
-        a.typ
+        PArray(coerce[PArray](a.pType).elementType, a.pType.required)
       case ArrayFlatMap(a, name, body) =>
-        TArray(coerce[TContainer](body.typ).elementType)
+        PArray(coerce[PContainer](body.pType).elementType, a.pType.required)
       case ArrayFold(a, zero, accumName, valueName, body) =>
-        assert(body.typ == zero.typ)
-        zero.typ
+        assert(body.pType == zero.pType)
+        zero.pType
       case ArrayScan(a, zero, accumName, valueName, body) =>
-        assert(body.typ == zero.typ)
-        TArray(zero.typ)
+        assert(body.pType == zero.pType)
+        PArray(zero.pType)
       case AggFilter(_, aggIR) =>
-        aggIR.typ
+        aggIR.pType
       case AggExplode(array, name, aggBody) =>
-        aggBody.typ
+        aggBody.pType
       case AggGroupBy(key, aggIR) =>
-        TDict(key.typ, aggIR.typ)
+        PDict(PType.canonical(key.pType), aggIR.pType)
```

Comment:
Why the canonical here?  Since extract aggregators rewrites this, I almost feel like it doesn't matter.

Reply:
ah, OK. But canonical because we're serializing to safe values and back in some cases.